src/contracts/MslV4.ts

```diff
@@ -236,17 +236,51 @@ export class MslV4 extends BaseContract<typeof multiSourceLoanABIV4> {
     return 0;
   }
 
-  async refinanceFullLoan({
-    offer,
-    signature,
-    loan,
+  async refinanceBatch({
+    renegotiationId,
+    refinancings,
   }: {
-    offer: RenegotiationV4;
-    signature: Hash;
-    loan: LoanV4;
+    renegotiationId: number;
+    refinancings: {
+      loan: LoanV4;
+      newAprBps: bigint;
+      sources: {
+        source: LoanV4['source'][number];
+        refinancingPrincipal: bigint;
+      }[];
+    }[];
   }) {
-    const txHash = await this.safeContractWrite.refinanceFull([offer, loan, signature]);
+    const offers: RenegotiationV4[] = [];
+    const loans: LoanV4[] = [];
+
+    // Generate (renegotiation offer, loan) pairs
+    refinancings.forEach(({ loan, sources, newAprBps }, index) => {
+      const targetPrincipal = loan.source.map(({ principalAmount, loanId }) => {
+        const refinancingSource = sources.find(({ source }) => source.loanId === loanId);
+        return refinancingSource ? principalAmount - refinancingSource.refinancingPrincipal : 0n;
+      });
+      const refinancingPrincipalAmount = sources.reduce(
+        (acc, { refinancingPrincipal }) => acc + refinancingPrincipal,
+        0n,
+      );
```

Comment:
I would move that sum to a helper function
```suggestion
      const refinancingPrincipalAmount = sum(sources.map({refinancingPrincipal} => refinancingPrincipal))
```

```suggestion
      const refinancingPrincipalAmount = sumBy(sources, 'refinancingPrincipal')
```

Reply:
https://github.com/gondixyz/gondi-js/pull/65/commits/e347f5f9a259169c48335a3a1dc4d4dcc0121816